File: frontend/src/app/core/services/service-type.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ServiceType, ServiceCategory, ServiceTypeRequest } from '../models';

@Injectable({ providedIn: 'root' })
export class ServiceTypeService {
  private readonly apiUrl = `${environment.apiUrl}/services`;

  constructor(private http: HttpClient) {}

  getServices(category?: ServiceCategory): Observable<ServiceType[]> {
    const params = category ? { category } : {};
    return this.http.get<ServiceType[]>(this.apiUrl, { params });
  }

  getService(serviceId: string): Observable<ServiceType> {
    return this.http.get<ServiceType>(`${this.apiUrl}/${serviceId}`);
  }

  // --- Admin ---

  getAllServices(): Observable<ServiceType[]> {
    return this.http.get<ServiceType[]>(`${this.apiUrl}/admin/all`);
  }

  createService(request: ServiceTypeRequest): Observable<ServiceType> {
    return this.http.post<ServiceType>(this.apiUrl, request);
  }

  updateService(
    serviceId: string,
    request: ServiceTypeRequest
  ): Observable<ServiceType> {
    return this.http.put<ServiceType>(`${this.apiUrl}/${serviceId}`, request);
  }

  deleteService(serviceId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${serviceId}`);
  }
}

File: backend/pawforward-api/src/main/java/com/pawforward/api/service/ServiceTypeService.java
package com.pawforward.api.service;

import com.pawforward.api.service.dto.ServiceTypeRequest;
import com.pawforward.api.service.dto.ServiceTypeResponse;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
public class ServiceTypeService {

    private final ServiceTypeRepository serviceTypeRepository;

    public ServiceTypeService(ServiceTypeRepository serviceTypeRepository) {
        this.serviceTypeRepository = serviceTypeRepository;
    }

    @Transactional(readOnly = true)
    public List<ServiceTypeResponse> getAllActiveServices() {
        return serviceTypeRepository.findByActiveTrueOrderBySortOrder().stream()
                .map(ServiceTypeResponse::from)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<ServiceTypeResponse> getServicesByCategory(ServiceCategory category) {
        return serviceTypeRepository.findByCategoryAndActiveTrueOrderBySortOrder(category).stream()
                .map(ServiceTypeResponse::from)
                .toList();
    }

    @Transactional(readOnly = true)
    public ServiceTypeResponse getService(UUID serviceId) {
        ServiceType service = findById(serviceId);
        return ServiceTypeResponse.from(service);
    }

    @Transactional(readOnly = true)
    public List<ServiceTypeResponse> getAllServices() {
        return serviceTypeRepository.findAll().stream()
                .map(ServiceTypeResponse::from)
                .toList();
    }

    @Transactional
    public ServiceTypeResponse createService(ServiceTypeRequest request) {
        ServiceType service = ServiceType.builder()
                .name(request.getName().trim())
                .category(request.getCategory())
                .description(request.getDescription())
                .durationMinutes(request.getDurationMinutes())
                .maxParticipants(request.getMaxParticipants())
                .price(request.getPrice())
                .depositAmount(request.getDepositAmount())
                .active(request.getActive() != null ? request.getActive() : true)
                .sortOrder(request.getSortOrder() != null ? request.getSortOrder() : 0)
                .imageUrl(request.getImageUrl())
                .build();

        service = serviceTypeRepository.save(service);
        return ServiceTypeResponse.from(service);
    }

    @Transactional
    public ServiceTypeResponse updateService(UUID serviceId, ServiceTypeRequest request) {
        ServiceType service = findById(serviceId);

        service.setName(request.getName().trim());
        service.setCategory(request.getCategory());
        service.setDescription(request.getDescription());
        service.setDurationMinutes(request.getDurationMinutes());
        service.setMaxParticipants(request.getMaxParticipants());
        service.setPrice(request.getPrice());
        service.setDepositAmount(request.getDepositAmount());
        if (request.getActive() != null) {
            service.setActive(request.getActive());
        }
        if (request.getSortOrder() != null) {
            service.setSortOrder(request.getSortOrder());
        }
        service.setImageUrl(request.getImageUrl());

        service = serviceTypeRepository.save(service);
        return ServiceTypeResponse.from(service);
    }

    @Transactional
    public void deleteService(UUID serviceId) {
        ServiceType service = findById(serviceId);
        serviceTypeRepository.delete(service);
    }

    private ServiceType findById(UUID serviceId) {
        return serviceTypeRepository.findById(serviceId)
                .orElseThrow(() -> new EntityNotFoundException("Service not found"));
    }
}
